File: 2026/06-June/26-Friday/solution.cs
public class BIT
{
    private int[] tree;
    private int n;

    public BIT(int n)
    {
        this.n = n;
        tree = new int[n + 1];
    }

    public void Update(int idx, int val)
    {
        while (idx <= n)
        {
            tree[idx] += val;
            idx += idx & -idx;
        }
    }

    public int Query(int idx)
    {
        int sum = 0;
        while (idx > 0)
        {
            sum += tree[idx];
            idx -= idx & -idx;
        }
        return sum;
    }
}

public class Solution
{
    public long CountMajoritySubarrays(int[] nums, int target)
    {
        int n = nums.Length;
        BIT bit = new BIT(2 * n + 2);

        int prefix = n + 1;
        bit.Update(prefix, 1);

        long ans = 0;

        foreach (int x in nums)
        {
            prefix += (x == target) ? 1 : -1;
            ans += bit.Query(prefix - 1);
            bit.Update(prefix, 1);
        }

        return ans;
    }
}

File: 2026/06-June/26-Friday/solution.py
class BIT:
    def __init__(self, n):
        self.n = n
        self.bit = [0] * (n + 1)

    def update(self, idx, val):
        while idx <= self.n:
            self.bit[idx] += val
            idx += idx & -idx

    def query(self, idx):
        s = 0
        while idx > 0:
            s += self.bit[idx]
            idx -= idx & -idx
        return s


class Solution:
    def countMajoritySubarrays(self, nums, target):
        n = len(nums)
        bit = BIT(2 * n + 2)

        prefix = n + 1
        bit.update(prefix, 1)

        ans = 0

        for x in nums:
            prefix += 1 if x == target else -1
            ans += bit.query(prefix - 1)
            bit.update(prefix, 1)

        return ans
